pmlc/rt/device.h

```diff
@@ -0,0 +1,47 @@
+// Copyright 2020, Intel Corporation
+
+#pragma once
+
+#include <memory>
+#include <stdexcept>
+#include <utility>
+
+#include "pmlc/rt/executable.h"
+
+namespace pmlc::rt {
+
+// Device represents a PlaidML device, capable of evaluating a PlaidML program.
+class Device {
+public:
+  // Returns the current Device, local to the current thread.
+  //
+  // If there is no current device, or the current device is of an unexpected
+  // type, an exception will be thrown.
+  //
+  // This is used to smuggle the current device through to runtimes that are
+  // directly invoked by lowered code, when that lowered code doesn't support
+  // the use of a passthrough parameter.
+  template <typename T>
+  static std::shared_ptr<T> current() {
```

Comment:
I thought we were going to pass the device pointer as a void* into the JIT'd code. If so, why do we need this?

Reply:
(Actually -- not the device pointer, but an invocation pointer.)